### PR TITLE
HHH-16521 Regression with target type in CAST function

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-16521

### CODE REVIEW

File: hibernate-core/src/test/java/org/hibernate/orm/test/nationalized/CastToNationalizedStringTest.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+		);
+	}
+
+	@Test
+	public void testQuery(SessionFactoryScope scope) {
+		final SQLStatementInspector sqlStatementInspector = (SQLStatementInspector) scope.getStatementInspector();
+		sqlStatementInspector.clear();
+		scope.inTransaction(
+				session -> {
+					session.createQuery( "select CAST(t.name AS nstring(200)) from TestEntity t" )
```
Comment (gavinking) [inline hibernate-core/src/test/java/org/hibernate/orm/test/nationalized/CastToNationalizedStringTest.java:53]: But is this legal HQL?? Where do we document that you're allowed to write this?

In HQL 6, the target type of a `cast()` is supposed to be an unqualified Java class name:

<https://docs.jboss.org/hibernate/orm/6.4/querylanguage/html_single/Hibernate_Query_Language.html#function-cast>

**Comment by gavinking** [comment]
I don't think I like this approach.

How about if instead the `cast()` function just respected the setting `hibernate.use_nationalized_character_data`?

So that you would write `cast(whatever as String)`, and it would automatically generate a SQL cast to `nvarchar`?

**Comment by dreab8** [comment]
I'm a little torn about `use_nationalized_character_data` influencing the `cast`, I'm thinking about the cast to nstring and I realized I'm not fully understanding the use case :(

**Comment by gavinking** [comment]
Yeah good point, let's try to understand the use case first 👍

**Pull request closed without merging**